feature/select

thatsamsonkid opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our
      guidelines: https://github.com/goetzrobin/spartan/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

- [ ] accordion
- [ ] alert
- [ ] alert-dialog
- [ ] aspect-ratio
- [ ] avatar
- [ ] badge
- [ ] button
- [ ] calendar
- [ ] card
- [ ] checkbox
- [ ] collapsible
- [ ] combobox
- [ ] command
- [ ] context-menu
- [ ] data-table
- [ ] date-picker
- [ ] dialog
- [ ] dropdown-menu
- [ ] hover-card
- [ ] input
- [ ] label
- [ ] menubar
- [ ] navigation-menu
- [ ] popover
- [ ] progress
- [ ] radio-group
- [ ] scroll-area
- [X] select
- [ ] separator
- [ ] sheet
- [ ] skeleton
- [ ] slider
- [ ] switch
- [ ] table
- [ ] tabs
- [ ] textarea
- [ ] toast
- [ ] toggle
- [ ] tooltip
- [ ] typography

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Closes #

Select Primitive

Closes #65 

## What is the new behavior?

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

goetzrobin commented:
Awesome! This is a great start! How deep do you want the feedback to go? Just overall if this is going into the right direction or is there any more specific things you'd like me to address?

Thanks for working on this! I am super excited for this!

thatsamsonkid commented:
I think overall/high-level for now. Im still tidying up everything. Can definitely do a more in depth one once I think this is closer to being done. 

I think one thing I also wanted an opinion on is do you think we should have stories for just brain usage and one with only hlm? 

1. I think to make sure someone can reasonable recreate everything with just brain

2.to make sure hlm is exposing everything it should from underlying brain and just using brain correctly overall? What do you think?

goetzrobin commented:
So one thing we decided to do going forward is to allow him to depend on brain directly. This means him can use brain components in their own template or via host directives, etc.

However, brain should never depend on helm and we should expose enough information so you can build your own UI lib just with brain components. 

Radix UI = Brain
shadcn = Helm

We are implementing both at the same time so we need to be a little more careful, but overall brain should do all the work and him only does some minor additions, adds styles, and improves API surface for those using them together!


Overall, I really like how this is coming along though! I think you are absolutely on the right track!

goetzrobin commented:
@thatsamsonkid let me know if I can do anything else to help or if you got everything you need for this! This is going to be a great addition 👍

thatsamsonkid commented:
Definitely!, I think I was getting hung up on some things for it and just needed a small break from it 😅. Let me verify I pushed everything I have so far and maybe I can get help with some specific things

thatsamsonkid commented:
Ok I updated this branch with latest and wanted to give an overview of what is remaining for this

So remaining items/issues:

- Determining minimum number of items to show. Should we have a default. Configurable number. Im a little unsure how radix/shadcn is determining whether to show few options or if you see their "scrolling" example its showing almost as many as possible it seems.But seemingly without any like direct input - need some consensus on this should work i guess

- Need to fix overlay cdk to open above or below depending on space - could use assistance on this or an example could help me also

- Showing and hiding the scrolling arrows needs to be fixed mainly an issue with where and how Im setting the dropdown scroll

- Tests for Select

- Minor issue with selected option showing as checked. Is only updating after hovering over old option

@goetzrobin

goetzrobin commented:
Sweet, I'll try to check this out this week! @elite-benni feel free to also take a look at this if you have time :)

theo-matzavinos commented on libs/ui/select/brain/src/index.ts:
```diff
@@ -0,0 +1,26 @@
+export * from './lib/brn-select-trigger.directive';
+export * from './lib/brn-select.component';
+export * from './lib/brn-select.service';
+
```
Missing `export * from './lib/brn-select-value.component';`

thatsamsonkid replied:
Sorry on the late response but good catch, thanks!

elite-benni commented on libs/ui/select/brain/src/lib/brn-select-trigger.directive.ts:
```diff
@@ -0,0 +1,38 @@
+	},
+})
+export class BrnSelectTriggerDirective implements OnDestroy {
+	unsubscribe = new Subject<boolean>();
```
unused code?

thatsamsonkid replied:
I think I was preparing to subscribe to something but then just didn't end up doing it that way haha but yes good eyes @elite-benni

elite-benni reviewed (changes requested):
Looks pretty good already. Left some comments.

goetzrobin commented on libs/ui/select/brain/src/lib/brn-select-content.component.ts:
```diff
@@ -0,0 +1,110 @@
+	host: {
+		'[attr.aria-labelledBy]': 'labelledBy()',
+		'[attr.aria-controlledBy]': "id() +'-trigger'",
+		class: 'flex flex-col',
```
Agreed that the tailwind classes are an issue here... While these two classes are probably used in 99.99% of TW projects we shouldn't rely on them...

thatsamsonkid replied:
Yea definitely understood, I did not initially realize brn should also work outside of tailwind! I switch these to inline styles!

goetzrobin commented on libs/ui/select/brain/src/lib/brn-select-content.component.ts:
```diff
@@ -0,0 +1,110 @@
+@Component({
+	selector: 'brn-select-content, hlm-select-content:not(noHlm)',
+	standalone: true,
+	imports: [BrnSelectScrollUpDirective, BrnSelectScrollDownDirective, HlmIconComponent],
```
Agreed! Maybe we can work with content projection here?

elite-benni replied:
That is the same problem, that I had with the switch.
When using content projection, the api is not as nice as it would be now. :/
It is already possible to use content projection, hlm Icon is used if no brnSelectScrollUp or brnSelectScrollDown is provided in the content.
So technically this is not a "tight" dependency, it would also work without Hlm.

thatsamsonkid replied:
Hmmm I had originally imagined that there would be a default icon and you would only need to provide the up/down icon arrows through projection if you wished to change the default to some other icon or icon family.  I see two potential options:

1. A brnIcon of some sort (would this just be ngIcon lol) - probably would not be different from current hlm really and would of course make ng-icon a dep for this component so probably not the best idea but an idea none the less.

2. We would have to require the user to project both and up and down icon and maybe throw some warning if they don't (provide both that is) and maybe just keep a placeholder text until hlmIcon contents are provided and or simply just keep this as an optional thing. If you don't provide content we can just leave this "hover scrolling areas" out.

Let me know what you guys think or open to any other suggestions also

elite-benni replied:
For me 2. sounds like the best of these options. 
The other option would be to make an helm component that uses the correct content, but this has the disadvantage of adding a new layer again. :/

thatsamsonkid replied:
Yea I think lets just keep it as optional projections for these "hover scroll areas" optional. I'll make that change and i think just add some console warning if only one is provided not the other.

goetzrobin commented:
Agree with @elite-benni's comments! I'll need to find some time to give this the attention it deserves! Thanks for your great work on this @thatsamsonkid! Probably one of the hardest and most impactful primitives! I think we are pretty close on this!!

thatsamsonkid commented on libs/ui/select/brain/src/lib/brn-select-scroll-down.directive.ts:
```diff
@@ -0,0 +1,11 @@
+	standalone: true,
+	host: {
+		'aria-hidden': 'true',
+		class: 'flex cursor-default items-center justify-center py-1',
```
I may not need these depending on our decision on the up and down arrows. But yes will definitely remove the tw classes

thatsamsonkid commented on libs/ui/select/helm/src/lib/hlm-select-content.directive.ts:
```diff
@@ -0,0 +1,29 @@
+})
+export class HlmSelectContentDirective {
+	baseClasses =
+		'top-[2px] relative z-50 min-w-[8rem] overflow-scroll rounded-md border bg-popover text-popover-foreground shadow-md p-1 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right] ...
```
Hawk Eyes @elite-benni they call him
<img src="https://media4.giphy.com/media/ihYG12Kt2mQf1LWB78/giphy.gif"/>

thatsamsonkid commented:
Alrighty update on changes, seems never ending for some reason 🥲

Completed Items/fixes:
- [x] Showing and hiding the scrolling arrows needs to be fixed mainly an issue with where and how Im setting the dropdown scroll 

- [x] Minor issue with selected option showing as checked. Is only updating after hovering over old option

- [x] Fixed the selections so if multiple selections are made and mutilple value goes from true -> false then selections are cleared. (Turns out cdkListbox does this by default as well so we're on the right track there)

So remaining items/issues:

- [ ] still need some input on how we should determine/configure number of items to show at once by select, not only for default but also for scrollable.

- [ ] Need to fix overlay cdk to open above or below depending on space - could use assistance on this or an example could help me also

- [ ] While I fixed the up/down arrow show hide and made some select "viewport" updates, still needs to be tweaked slightly with up/down arrows covering very first and last options

- [ ] In dark mode the select dropdown out line is a bright white. Not entirely sure why

- [ ] Group heading and section directive styling still needs to be worked on

- [ ] Optional group heading sticky 

- [ ] Tests for Select

@goetzrobin @elite-benni

thatsamsonkid commented:
https://github.com/goetzrobin/spartan/assets/22568206/ee4b3134-b073-4487-8603-438b3e5ca907

👀👀👀

elite-benni commented on libs/ui/select/brain/src/lib/brn-select-group.directive.ts:
```diff
@@ -1,21 +1,13 @@
-		return this._labelledBy();
-	}
-	readonly _labelledBy = signal('');
+	labelledBy = signal('');
```
was this supposed to be an input?

thatsamsonkid commented on apps/app/src/app/pages/(components)/components/(select)/select--multiple.preview.ts:
```diff
@@ -0,0 +1,69 @@
+	imports: [ReactiveFormsModule, BrnSelectImports, HlmSelectImports],
+	providers: [provideIcons({ radixChevronUp, radixChevronDown })],
+	template: `
+		<form [formGroup]="options">
```
Yes please

goetzrobin commented:
@thatsamsonkid anything I can help to get this into main? Keep up the awesome work!

thatsamsonkid commented:
I should have some time tomorrow to hopefully wrap up the tests.

Mainly just need to add tests for some of the keyboard actions and scrolling. Then refine docs maybe. I was thinking I should probably add a few useful event outputs like on open/close to start with.

I think autofill is still an issue here but I think I will resolve that once I help convert combobox and command to our own implementation since its very similar.

@goetzrobin

thatsamsonkid commented:
All, I think this might be ready for serious consideration to merge finally!. Do we need to test cli/nx setup? I think docs we can do more but I know we were still looking into how to best format inputs/outputs of the components. 

But for now the same level of docs seems to be there now thanks @elite-benni and I think I got a good number of tests there to test the core functionality!

Let me know if anything

@goetzrobin @elite-benni

thatsamsonkid commented on apps/app/src/styles.css:
```diff
@@ -198,3 +198,7 @@ select {
+
+* {
+	border-color: hsl(var(--border));
+}
```
@goetzrobin @elite-benni - just wanted to call special attention to this change. Im pretty sure I noticed this was missing in storybook and saw there was a similar issue in the docs too with the select component border. When I checked shadcn I found this was missing but might just want to double check all the components make sure no adverse effects between dark/light modes. At quick glance everything seems fine
